File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/handler/websocket/FetchConnectionsRequestHandler.java
package net.prostars.messagesystem.handler.websocket;

import java.util.List;
import net.prostars.messagesystem.constant.IdKey;
import net.prostars.messagesystem.dto.domain.Connection;
import net.prostars.messagesystem.dto.domain.UserId;
import net.prostars.messagesystem.dto.websocket.inbound.FetchConnectionsRequest;
import net.prostars.messagesystem.dto.websocket.outbound.FetchConnectionsResponse;
import net.prostars.messagesystem.service.ClientNotificationService;
import net.prostars.messagesystem.service.UserConnectionService;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings("unused")
public class FetchConnectionsRequestHandler implements BaseRequestHandler<FetchConnectionsRequest> {

  private final UserConnectionService userConnectionService;
  private final ClientNotificationService clientNotificationService;

  public FetchConnectionsRequestHandler(
      UserConnectionService userConnectionService,
      ClientNotificationService clientNotificationService) {
    this.userConnectionService = userConnectionService;
    this.clientNotificationService = clientNotificationService;
  }

  @Override
  public void handleRequest(WebSocketSession senderSession, FetchConnectionsRequest request) {
    UserId senderUserId = (UserId) senderSession.getAttributes().get(IdKey.USER_ID.getValue());
    List<Connection> connections =
        userConnectionService.getUsersByStatus(senderUserId, request.getStatus()).stream()
            .map(user -> new Connection(user.username(), request.getStatus()))
            .toList();
    clientNotificationService.sendMessage(
        senderSession, senderUserId, new FetchConnectionsResponse(connections));
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/FetchConnectionsRequestHandler.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.constant.IdKey
import com.yscorp.withpush.messagesystem.dto.domain.Connection
import com.yscorp.withpush.messagesystem.dto.domain.UserId
import com.yscorp.withpush.messagesystem.dto.websocket.inbound.FetchConnectionsRequest
import com.yscorp.withpush.messagesystem.dto.websocket.outbound.FetchConnectionsResponse
import com.yscorp.withpush.messagesystem.service.ClientNotificationService
import com.yscorp.withpush.messagesystem.service.UserConnectionService
import net.prostars.messagesystem.constant.IdKey
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession

@Component
@Suppress("unused")
class FetchConnectionsRequestHandler(
    private val userConnectionService: UserConnectionService,
    private val clientNotificationService: ClientNotificationService
) : BaseRequestHandler<FetchConnectionsRequest> {

    /**
     * 사용자의 특정 상태(ACCEPTED, PENDING 등)에 해당하는 유저 목록을 조회하여
     * Connection 객체로 변환한 후 응답 메시지로 전송합니다.
     */
    override fun handleRequest(senderSession: WebSocketSession, request: FetchConnectionsRequest) {
        val senderUserId = senderSession.attributes[IdKey.USER_ID.value] as? UserId ?: return

        val status = request.status
        val connections = userConnectionService.getUsersByStatus(senderUserId, status)
            .map { user -> Connection(user.username, status) }

        clientNotificationService.sendMessage(
            senderSession,
            senderUserId,
            FetchConnectionsResponse(connections)
        )
    }
}
